src/libserver/session.cpp

```diff
@@ -1684,24 +1688,48 @@ JsonApiResult Session::callListingsJsonApi(
 	JsonApiMethod method, const QStringList &path, const QJsonObject &request)
 {
 	Q_UNUSED(request);
-	if(path.length() != 1)
+
+	int pathLength = path.length();
+
+	if(pathLength > 1)
 		return JsonApiNotFound();
-	const int id = path.at(0).toInt();
 
-	const auto announcements = m_announcements->getAnnouncements(this);
-	for(const sessionlisting::Announcement &a : announcements) {
-		if(a.listingId == id) {
-			if(method == JsonApiMethod::Delete) {
-				unlistAnnouncement(a.apiUrl.toString());
-				return JsonApiResult{
-					JsonApiResult::Ok,
-					QJsonDocument(QJsonObject{{"status", "ok "}})};
+	if(pathLength == 1) {
+		const int id = path.at(0).toInt();
 
-			} else {
-				return JsonApiBadMethod();
+		const auto announcements = m_announcements->getAnnouncements(this);
+		for(const sessionlisting::Announcement &a : announcements) {
+			if(a.listingId == id) {
+				if(method == JsonApiMethod::Delete) {
+					unlistAnnouncement(a.apiUrl.toString());
+					return JsonApiResult{
+						JsonApiResult::Ok,
+						QJsonDocument(QJsonObject{{"status", "ok "}})};
+
+				} else {
+					return JsonApiBadMethod();
+				}
 			}
 		}
+
+		return JsonApiNotFound();
+	}
+
+	if (method == JsonApiMethod::Get) {
+		return JsonApiResult{
+			JsonApiResult::Ok, QJsonDocument(getListingsDescription())};
+			
+	} else if(method == JsonApiMethod::Create) {
+		if(request.contains("url")) {
+			makeAnnouncement(QUrl(request["url"].toString()));
```

Comment:
The `makeAnnouncement` function doesn't behave ideally for this. It'll send a message into the session if listing is not allowed and always returns `void` instead of some error indicator, so the API caller won't notice the error. I think it's okay to leave it like that, since it's unlikely to cause problems in practice and it's not like the web UI gives proper error feedback anyway, but maybe add a comment about it here.

Reply:
It'd need to touch more code to make it possible since the `Announcements` doesn't offer other ways. Let's leave it like that for now.